netdata/netdata
# Add option for eBPF plugin to manipulate memory

##### Summary
This PR is bringing another method to adjust the memory usage for eBPF.plugin. This new method will be set by default, because it needs less iteration with the kernel, on the other hand it will use more memory from computer. My initial conclusion is that it worth while to use more memory and speed up process and environment. We are also adding a new parameter that allows user to control the memory mode.

![fix](https://user-images.githubusercontent.com/49162938/165199780-4c9476b7-644c-492d-b688-77c2d39ed8be.png)


##### Test Plan

1. Update your `/etc/netdata/ebpf.d.conf` adding the new parameter:
```conf
[global]
    allocate memory = static
```
2. Clone this branch and compile it.
3. Run Netdata and you will see that eBPF.plugin will have memory and cpu usage smaller than `apps`. Depending of your kernel version, the same problems reported [here](https://github.com/netdata/netdata/issues/12719#issuecomment-1112808123) will happen.
4. Run `curl -k -o FILE.txt "https://localhost:19999/api/v1/data?chart=apps.mem&dimension=ebpf.plugin"` after few minutes, and you will see a "stable" `memory`.
5. Stop netdata, and set adjust your `/etc/netdata/ebpf.d.conf`:
```conf
[global]
    allocate memory = dynamic
```
6. Run step 4 again, and values won't be so stable.

We expect to have data collection normally with both options. Of course with `dynamic` we expect major iteration with kernel and more CPU and memory usage.

##### Additional Information
It is not necessary to test on different kernels, because we are changing only `user ring` code.

The following table demonstrate the results using current `master` branch and this PR:

| Branch | Result |
|------------|----------|
| `master` | [old.txt](https://github.com/netdata/netdata/files/8602063/old.txt) |
| This PR |  [new.txt](https://github.com/netdata/netdata/files/8602065/new.txt)|


<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? Users will have less memory and cpu usage for `ebpf.plugin`
- How is the user impacted by the change?  His machine will have more resources for other tasks.
- What are there any benefits of the change?  we are improving data collection
</details>
